packages/service/src/plugin.ts

```diff
@@ -26,18 +43,69 @@ export interface PluginLogger {
 }
 
 export class PluginsRunner {
-  plugins: { plugin: Plugin<PluginOptions>; options: PluginOptions }[] = [];
+  levels: {
+    plugin: PluginFactory;
+    options?: PluginOptions;
+    filter?: (fileName: string) => boolean;
+  }[] = [];
   context: PluginsRunnerContext;
 
   constructor(context: PluginsRunnerContext) {
     this.context = context;
   }
 
-  queue<P extends Plugin<PluginOptions>>(
-    plugin: P,
-    options: P extends Plugin<infer O> ? O : never
+  queue<
+    PluginType extends Plugin = Plugin,
+    PluginOptionsType extends PluginOptions = PluginOptions
+  >(plugin: PluginFactory<PluginType, PluginOptionsType>): this;
+  queue<
+    PluginType extends Plugin = Plugin,
+    PluginOptionsType extends PluginOptions = PluginOptions
+  >(
+    plugin: PluginFactory<PluginType, PluginOptionsType>,
+    filter: (fileName: string) => boolean
+  ): this;
+  queue<
+    PluginType extends Plugin = Plugin,
+    PluginOptionsType extends PluginOptions = PluginOptions
+  >(plugin: PluginFactory<PluginType, PluginOptionsType>, options: PluginOptionsType): this;
+  queue<
+    PluginType extends Plugin = Plugin,
+    PluginOptionsType extends PluginOptions = PluginOptions
+  >(
+    plugin: PluginFactory<PluginType, PluginOptionsType>,
+    options: PluginOptionsType,
+    filter: (fileName: string) => boolean
+  ): this;
+  queue<
+    PluginType extends Plugin = Plugin,
+    PluginOptionsType extends PluginOptions = PluginOptions
+  >(
+    plugin: PluginFactory<PluginType>,
+    ...optionsOrFilter:
+      | []
+      | [PluginOptionsType]
+      | [(fileName: string) => boolean]
+      | [PluginOptionsType, (fileName: string) => boolean]
```

Comment:
This is hard to read. There is a lot of overloading and typing. Please add some documentation why this is necessary or simplify if at all possible.